Unify search logic for compilers and externals

alalazo opened the pull request:
fixes #39627

This PR is a slight step back from #44419, to refactor code in a way that doesn't affect where we write compilers.

The code used for:
```console
$ spack compiler find
```
is the same as the one used for:
```console
$ spack external find
```
Dead code in `spack.compilers` has been removed, and more can be removed later from each compiler class.

Compared to #44419, compilers are still written in `compilers.yaml` - so users shouldn't be negatively affected by this change.

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

alalazo committed:
spack compiler find: write into packages.yaml

This commit makes spack compiler find write in packages.yaml, instead of
compilers.yaml.

alalazo committed:
Add -j option to spack compiler find

alalazo committed:
Remove detection test for compiler objects

alalazo committed:
Extend compiler find search paths for Windows

alalazo committed:
Filter valid compilers

This excludes, e.g. llvm~clang etc. if found

alalazo committed:
Recover mixed-toolchain options

alalazo committed:
Remove redundant test

add_compilers_to_config is not called anymore

alalazo committed:
Revert writing in packages.yaml by default

haampie commented:
`strace -cf` shows it's not equivalent yet.

1. regexes seem to include more executables (cross compilers which are then listed with an incorrect target)
2. executables are queried repeatedly/redundantly about their version using the same flag.

This makes it 2.5x slower for me.

alalazo committed:
Make sure we don't inspect the same dir twice under different names

alalazo commented:
The main issue for the slowdown you still see seems to be the overhead of loading a package (which we don't have on `develop`). This is a trace of a worker on my laptop:
![Screenshot from 2024-08-16 23-49-26](https://github.com/user-attachments/assets/d0f4aafa-cc60-4180-9ead-66bfaf449c9a)

The ratio for `llvm` seems to be 0.4 secs to load the pkg, 0.1 secs to detect the spec.

alalazo committed:
Fix -j argument to spack compiler find

haampie commented on lib/spack/spack/compilers/__init__.py:
```diff
@@ -380,79 +262,75 @@ def all_compiler_specs(scope=None, init_config=True):
     default_paths = fs.search_paths_for_executables(*path_hints)
+    if sys.platform == "win32":
+        default_paths.extend(windows_os.WindowsOs().compiler_search_paths)
+    compiler_pkgs = spack.repo.PATH.packages_with_tags(COMPILER_TAG, full=True)
```
import

haampie commented on lib/spack/spack/compilers/__init__.py:
```diff
@@ -380,79 +262,75 @@ def all_compiler_specs(scope=None, init_config=True):
-    return make_compiler_list(
-        [remove_errors(detected) for detected in detected_versions if valid_version(detected)],
-        mixed_toolchain=mixed_toolchain,
+    detected_packages = spack.detection.by_path(
```
same

haampie commented:
ThreadpoolExecutor works fine and is significantly faster: 3s -> 0.5s on macOS

alalazo commented:
> ThreadpoolExecutor worked and works fine and is significantly faster: 3s -> 0.5s on macOS. I guess it's relatively safe although we don't control the package specific implementation... but that's unlikely to have global side effects.

Seems there's some work to do to use a ThreadPoolExecutor:
```console
$ spack external find -t compiler
==> Warning: error detecting "llvm" from prefix /bin [trying to set variants "clang, lld, or lldb" in package "llvm", but the package has no such variants [happened when validating 'llvm@10.0.0+clang~lld~lldb']]
==> Warning: error detecting "llvm" from prefix /usr/bin [trying to set variants "clang, lld, or lldb" in package "llvm", but the package has no such variants [happened when validating 'llvm@12.0.0+clang+lld~lldb']]
==> No new external packages detected
```
vs:
```console
$ spack external find -t compiler
==> The following specs have been detected on this system and added to /home/culpo/.spack/packages.yaml
gcc@8.4.0  gcc@9.4.0  gcc@10.5.0  llvm@10.0.0  llvm@11.0.0  llvm@12.0.0
```

alalazo committed:
Add missing imports, recovered previous format strings

alalazo committed:
Remove unneeded unit test for Apple Clang

We could restrict to test this only on Darwin, but it's
already tested in audits.
